src/resolve_reports

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+from collections import defaultdict
+import csv
+from os import environ
+
+from bs4 import BeautifulSoup
+
+from logic import (
+    determine_timestamp,
+    get_options,
+    get_date_from_filename,
+    uri_schema_doi,
+)
+from utils import doi_to_id, set_token_auth
+
+
+XREF_DOI_PREFIX = environ['XREF_DOI_PREFIX']
+
+
+ARGS = [
+    {
+        'val': '--measure',
+        'dest': 'measure',
+        'action': 'store',
+        'default': None,
+        'help': 'Measure URI, e.g. '
+                'https://metrics.operas-eu.org/citations/crossref/v1',
+        'required': True,
+    }, {
+        'val': '--add-headers',
+        'dest': 'add_headers',
+        'action': 'store_true',
+        'default': [],
+        'help': 'Whether to print report headers',
+        'required': False,
+    }, {
+        'val': '--in_file',
+        'dest': 'in_file',
+        'action': 'store',
+        'help': 'File to read Crossref citation data from (XML format).',
+        'required': True,
+    }, {
+        'val': '--out_file',
+        'dest': 'out_file',
+        'action': 'store',
+        'help': 'CSV file to save output to (CSV format).',
+        'required': True,
+    }
+]
+
+HEADER = (
+    'measure_id',
+    'timestamp',
+    'work_uri',
+    'country_uri',
+    'event_uri',
+    'value',
+)
+
+
+def get_crossref_citations(cached_file):
+    """Extract raw data from Crossref citation XML for correct DOIs only.
+
+    Args:
+        cached_file (str): Name of XML file containing data from Crossref.
+
+    Returns:
+        dict: Each DOI that is cited and the citation entries for that DOI.
+    """
+
+    with open(cached_file, 'r') as f:
+        xml_data = BeautifulSoup(f.read(), 'xml')
+
+    relevant_entries = filter(
+        lambda x: x.attrs.get('doi', '').startswith(XREF_DOI_PREFIX),
+        xml_data.find_all('forward_link')
+    )
+
+    citations = defaultdict(list)
+    for entry in relevant_entries:
+        doi = entry.attrs.get('doi')
+        citations[doi].extend(entry.find_all('journal_cite'))
+        citations[doi].extend(entry.find_all('book_cite'))
+
+    return citations
+
+
+def resolve(cached_file):
+    """Extract citation information (for metrics) from Crossref XML.
+
+    Args:
+        cached_file (str): Name of XML file containing data from Crossref.
+
+    Returns:
+        Generator: Yields a tuple with citation data.
+    """
+
+    file_date = get_date_from_filename(cached_file)
+    citations = get_crossref_citations(cached_file)
+
+    if citations:
+        set_token_auth()
+
+    for doi, entries in citations.items():
```

Comment:
You had better be sure `citations` is a `dict`

Reply:
Yes, `get_crossref_citations` should always return a `dict` :+1:

Reply:
Good